### ISSUE TITLE
GraphQL Bridge

### ISSUE THREAD

**Issue by theodorDiaconu** [cult-of-coders/grapher#235]
## GraphQL Bridge Ideas

Purpose is to create a set of tools that enable having very powerful backend APIs using Grapher. Helping with collection linking and also highly performant fetches.

## AST Manipulation (priority)

This way, with Grapher you avoid completely creating resolvers for the queries you need, because Grapher knows how to deal with it, or you can do something like. Simpler:

```js
// and what we need to do is something like
Users.setFieldMap({
   createdAt: 'created_at';
})
// if a fieldMap exists, grapher performs transformation in prepareForProcess() and prepareForDelivery()

```

For example, if you would like to restrict access to 'comments' link, you could do that using a directive inside GraphQL, but, you could also easily manipulate the body:

For fields that are different from the database we need a way to map-remap them. For that we will suggest to actually use addReducers functionality to achieve this. And maybe we could pass an option like avoid cleaning out stuff, since GraphQL will take care of that for us already, so we don't do additional processing.

```js
const query = Users.astToQuery(ast, {
   embody({body, getArgs}) {
       if (body.comments) {
            const args = getArgs('comments'); // should work nested too, getArgs('comments.user'), the args from the ast
            body.comments.$filters = {};
       }
   }
   $filters,
   $options,
   maxDepth, // in case you don't have an intersect
   maxLimit, // limit the maximum of results that are given no matter what
   allow, // only allow these fields/links/resolvers, restrict fields will make no sense, can also be a function
   restrict, // restricts only the top level fields <?>, can also be a function
   intersect: Body, // performs the full body intersection, should be done before applying selector, options
})
```

Allow/Restrictions must also work in depth like 'comments.user'

```js
setAstToQueryDefaults({
   maxLimit,
   maxDepth,
});
```

## Type Directives for Linking Collections (nice to have)

Each relation will be automatically indexed, and detect it's multiplicity. The API below is an alternative, but not so sure, maybe it's more elegant to have it separated.

```js
type User @mongo("users") {
    comments: [Comment] @link(to: "user")
}

type Comment @mongo("comments") {
   user: User @link(field: "userId")
   createdAt: Date @mapTo("created_at")
}
```
Field must always be specified from now on, and linking is now done automatically whether it's one or Many. What about meta links ?

After deep analysis and API exploration, metadata links will not be available for GraphQL querying, and they will be deprecated. 

Breakdown of tasks:
1. Allow to pass an option to Grapher so we don't do cleanups of leftovers
2. Create an AST to Grapher function that is properly tested

**Comment by theodorDiaconu** [comment]
Because named queries are so powerful we could query their subbody properly:
```js
query.fetch({body, params})
```

**Comment by theodorDiaconu** [comment]
Solved.
